[[ BytecodeBlocks ]] Add bytecode blocks

runrevmark opened the pull request:
This patch adds the ability to write bytecode directly inside 'bytecode'
blocks in LCB:

```
bytecode
  ...
end bytecode
```

Labels are local to the block and can be specified by using:

```
myLabel:
```

Temporaries are local to the block and can be specified by using:

   variable myTemp

Temporaries, local variables and parameters can be used as register type
arguments to opcodes.

Module variables, defined constants and handler identifiers can be used
as the 'source' parameter for the 'fetch' opcode.

Module variables can be used as the 'dest' parameter for the 'source'
opcode.

Handler identifiers can be used as the 'handler' parameter for the 'invoke'
opcode.

runrevmark commented:
This still needs a bit more work.

In particular, I need to separate out the autogen'd gentle methods into a separate file (which will still have to be run and updated manually - as we can't as yet use livecode script during the build process).

I'm not entirely happy with the way to declare temporaries in bytecode blocks but there are two things at the moment which have to happen first before it can change. The first is to make 'defaulting' of slots explicit in the bytecode; the second is to make variables block-scoped.

With these two changes, the variable clause inside a bytecode block will be the same as that outside of a bytecode block - although (as bytecode statements are absolutely minimal) they won't be allowed to have an explicit initialization clause on them, just a type.

Another thing I'm not entirely happy with is the bytecode emission API in libscript - I think it would be better to remove the specific methods in the API, and just have a generic 'emit bytecode' operation - the current individual methods could then be wrappers around that. This makes the API a cleaner, and also paves the way for supporting substantially more bytecodes which is on the cards.

runrevmark commented:
The bytecode block feature has now been restructured to work on information about the bytecode operations which is compiled into libscript - thus no additional information is needed and adding new bytecode operations will just requiring updating the internal data structure (at least as far as allowing the compiler to emit them... executing them is a different matter!).

There is currently a problem with compiling forward references, however, due to the way the generate module of the compiler works and the fact the bytecode emission API now does checking of the types of arguments.

Beyond that, this patch should probably wait until after block-scoped variables have been done so that the bytecode 'variable' definitions can be unified with handler 'variable' definitions - this can't be done at the moment as labels must have local scope, whilst variables have handler scope and identifiers are either block scoped or not!

peter-b commented:
> There is currently a problem with compiling forward references, however, due to the way the generate module of the compiler works and the fact the bytecode emission API now does checking of the types of arguments.

@runrevmark Not sure what you mean by "forward references" in this context -- could you please elaborate?

runrevmark committed:
[[ Compiler ]] Fix emit indicies on 64-bit

The code which generates indicies from libscript for constants was
passing a long to a uindex_t&.

runrevmark committed:
[[ NilAsUnassigned ]] Fix multi-invoke error case

When a multi-invoke fails to resolve it builds a list of the values
which it was selecting the handler on - including 'out' parameters.

Since unassigned slots are now nil and not kMCNull, it must map them
to kMCNull.

runrevmark committed:
[[ BytecodeBlocks ]] Add bytecode blocks

This patch adds the ability to write bytecode directly inside 'bytecode'
blocks in LCB:

    bytecode
      ...
    end bytecode

Labels are local to the block and can be specified by using:

    myLabel:

Temporaries are local to the block and can be specified by using:

   variable myTemp

Temporaries, local variables and parameters can be used as register type
arguments to opcodes.

Module variables, defined constants and handler identifiers can be used
as the 'source' parameter for the 'fetch' opcode.

Module variables can be used as the 'dest' parameter for the 'source'
opcode.

Handler identifiers can be used as the 'handler' parameter for the 'invoke'
opcode.

The bytecode information is recorded in a simple array in
script-builder.cpp - including name, opcode index and opcode
parameter format.

The script builder APIs have been thinned down to remove specific
bytecode emission functions. Instead, bytecode is emitted using
EmitBytecodeInModule. The emit module of lc-compile has been
updated to use this.

runrevmark commented:
@peter-b: When the compiler is generating code, it first generates declaration indicies which allow handlers declared after the call to be found. However, bytecode emission now does some checking of the arguments to an op at emission time, and to do that it needed the type of declaration. I've modified it now so when the declaration indicies are generated, it binds them to the kind of declaration as well, so the bytecode emission can do its checking.

peter-b commented:
@runrevmark I'm not sure that the fix for the multi-invoke error case is quite right.  Surely the multi-invoke should do a checked fetch and runtime error if the slot is unassigned?

peter-b commented:
@runrevmark I wonder if the bytecode would be more "visually distinct" from normal LCB code if it used `underscore_names` for opcodes instead of `TitleCaseNames`?

peter-b commented on libscript/src/script-builder.cpp:
```diff
@@ -22,6 +22,166 @@
+    const char *format;
+};
+
+static MCScriptBytecodeInfo s_bytecode_info[] =
```
This should probably be explicitly `const`, I think.

peter-b commented on libscript/src/script-builder.cpp:
```diff
@@ -1360,187 +1532,140 @@ void MCScriptResolveLabelForBytecodeInModule(MCScriptModuleBuilderRef self, uind
-void MCScriptContinueInvokeInModule(MCScriptModuleBuilderRef self, uindex_t p_arg_reg)
-{
-    if (self == nil || !self -> valid)
+    if (p_opcode >= sizeof(s_bytecode_info) / sizeof(s_bytecode_info[0]) ||
```
I think it might be a good idea to have an `inline` function that returns this value (i.e. number of known bytecodes), since it's used in several places.

peter-b commented on tests/lcb/compiler/bytecode.lcb:
```diff
@@ -0,0 +1,291 @@
+      Jump lAfterAssign
+      AssignConstant tTest, true
+   lAfterAssign:
+   end bytecode
```
While we're getting _mega_ bikesheddy: it's confusing that the label is "fully" unindented.  Maybe it would be clearer if it was indented by one space beyond the indentation of the `bytecode`/`end bytecode`?

peter-b commented on toolchain/lc-compile/src/grammar.g:
```diff
@@ -854,6 +859,34 @@
+        Identifier(-> Name) @(-> Position) ":"
+
+    'rule' Bytecode(-> variable(Position, Name, Type)):
+        "variable" @(-> Position) Identifier(-> Name) OptionalTypeClause(-> Type)
```
Would it make sense to call this `register` rather than `variable`?

runrevmark replied:
My original thinking here was that since variables are the same inside a bytecode block as outside - they are named, have a type, have block-scope - I figured it was suitable. However, there is a difference in that variables are not automatically 'reset' in bytecode context. This is probably enough justification that they should be described differently.

peter-b commented on toolchain/lc-compile/src/report.c:
```diff
@@ -27,13 +27,13 @@ extern int IsDependencyCompile(void);
 {
     s_error_count = 0;
-    s_verbose_level = 0;
+    s_verbose_level = 1;
```
Was changing the default verbosity level a deliberate change? It was deliberately set to 0 by default (and the `-v`/`--verbose` option was added) in order to avoid swamping the default build logs with compiler debugging output.

runrevmark replied:
That was an accident! I forgot to set it back to 0.

runrevmark commented:
In regards to the multi-invoke error then out parameters don't play a part in resolution - and they are the slots which might be nil, so we just convert to Null.

peter-b commented on docs/guides/LiveCode Builder Bytecode Reference.md:
```diff
@@ -0,0 +1,154 @@
+    assign_constant <register>, <constant>
+
+The assign_constant operation copies <constant> into <register>. Here <constant>
+can be 'nothing', a boolean, integer, real, string, list or array.
```
I assume that the `<constant>` needs to be a literal, right?

peter-b commented on docs/guides/LiveCode Builder Bytecode Reference.md:
```diff
@@ -0,0 +1,154 @@
+
+The invoke operation creates a new frame copying values from the argument registers
+for any in or inout parameters. It then starts executing the bytecode attached
+to <handler>.
```
Is the `<handler>` is permitted to be a register as well?

runrevmark replied:
No - it has to be a definition. 'invoke indirect' is the operation which loads the handler from a register

peter-b replied:
It might be worth mentioning that the `<handler>` needs to be a definition and that `<result>` is the register into which the return value from the invocation is placed (and the types need to match).

peter-b commented on docs/guides/LiveCode Builder Bytecode Reference.md:
```diff
@@ -0,0 +1,154 @@
+
+    assign_list <dst-reg>, <element1-reg>, ..., <elementn-reg>
+
+The assign_array operation builds a list value from <element1-reg> up to
```
"assign_array" should be "assign_list" here.

peter-b commented:
@livecode-vulcan review ok 629d60f

livecode-vulcan commented:
💙 review by @peter-b ok 629d60f7ab86ab52676e09819748c1a4db754615

livecode-vulcan commented:
😞 test failure 629d60f7ab86ab52676e09819748c1a4db754615
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/2615)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/2625)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/2610)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/2626)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/2632)
- try-community-win-x86: [failure](https://vulcan.livecode.com/builders/try-community-win-x86/builds/2638)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/2650)

runrevmark commented:
@livecode-vulcan retry 629d60f

livecode-vulcan commented:
😞 test failure 629d60f7ab86ab52676e09819748c1a4db754615
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/2617)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/2627)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/2612)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/2628)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/2634)
- try-community-win-x86: [failure](https://vulcan.livecode.com/builders/try-community-win-x86/builds/2640)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/2652)

peter-b commented:
@runrevmark Okay, I've read through the whole patch again and I have a few minor comments that probably don't strictly need to be addressed.  However, I'll include them for completeness!
- In the release note, I think you may need to indent the code sample by 4 spaces relative to the text of the bullet point in order for the markdown parser to process it correctly...
- Should `s_bytecode_{info,count}` be `k_` since they're constants?
- In `MCScriptEmitBytecodeInModuleA()`, the `kMCScriptByteCodeParameterTypeDefinition` case could probably be written more prettily using an inner `switch`. Actually, in general that function could be prettified if you put `__kind_of_definition(self, p_arguments[i])` into a local variable before entering the `switch(t_param_type)` block.
- Is `for(;;)` preferred over `while (true)` in LiveCode C++ style?
- I wonder if there should be a test for `reset` of a variable or register with a type that doesn't have a default value (but I'm not sure what the options are there)
- `s_type_map` in `EmitDefinitionIndex()` should probably be declared `const` (this gives some semantic protection but per discussion with @livecodefraser it won't change the generated code)
- Same for the `static __opcode_index` variables around the place that are never modified
- We should really do something about checking the result of memory allocations in lc-compile!
- Are the `MCLog()` statements in `BytecodeEnumerate()` left in intentionally?
- Is the `OPCODE` type in `types.g` still needed? I thought that the bytecode grammar was now generated

Notwithstanding these comments, this looks ✨ awesome✨ !
@livecode-vulcan review ok 3064160

livecode-vulcan commented:
💙 review by @peter-b ok 30641602015d34b2f3067c184454dd96d76c8339

livecode-vulcan commented:
😎 test success 30641602015d34b2f3067c184454dd96d76c8339
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/2621)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/2631)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/2616)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/2632)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/2638)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/2644)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/2656)

runrevmark committed:
[[ Cleanup ]] Cleanup various parts of bytecode block patch

This patch ensures static constants are constant and use appropriate
naming (k...).

This patch removes warnings from emit.cpp.

This patch tidies up MCScriptEmitBytecodeInModuleA

This patch removes unused code including the OPCODE definition and
MCLog()'s which should not have been committed.

peter-b commented:
@livecode-vulcan review ok 6a39fcf

livecode-vulcan commented:
💙 review by @peter-b ok 6a39fcfe66bfa31b3c11b40ddc81bf32df0cdc12

livecode-vulcan commented:
😎 test success 6a39fcfe66bfa31b3c11b40ddc81bf32df0cdc12
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/2625)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/2635)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/2620)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/2636)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/2642)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/2648)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/2660)
